Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSLoginWindowController.mm

```diff
@@ -0,0 +1,267 @@
+    __weak NSBlockOperation *weakOperation = operation;
+    [operation addExecutionBlock:^{
+        NetCliAuthAutoReconnectEnable(false);
+        
+        if (!NetCliAuthQueryConnected())
+            NetCommConnect();
+        NetCommAuthenticate(nullptr);
+        
+       while (!NetCommIsLoginComplete()) {
+           if(weakOperation.cancelled) {
+               return;
+           }
+           NetCommUpdate();
+       }
+       
+       ENetError result = NetCommGetAuthResult();
+       [NSOperationQueue.mainQueue addOperationWithBlock:^{
+           completion(result);
+       }];
+    }];
+    [_loginQueue addOperation:operation];
+}
+
+@end
+
+@implementation PLSLoginParameters
+
+static void *StatusTextDidChangeContext = &StatusTextDidChangeContext;
+
+- (id)init {
+    self = [super init];
+    [self load];
+    return self;
+}
+
+-(void)save {
+    //windows segments by product name here. in since user defaults belong to this product, we don't need to do that.
+    NSString *serverName = [NSString stringWithSTString:GetServerDisplayName()];
+    NSMutableDictionary *settingsDictionary = [[[NSUserDefaults standardUserDefaults] dictionaryForKey:serverName] mutableCopy];
+    if(!settingsDictionary)
+        settingsDictionary = [NSMutableDictionary dictionary];
+    [settingsDictionary setObject:self.username forKey:@"LastAccountName"];
+    [settingsDictionary setObject:[NSNumber numberWithBool:self.rememberPassword] forKey:@"RememberPassword"];
+    [[NSUserDefaults standardUserDefaults] setObject:settingsDictionary forKey:serverName];
+    [[NSUserDefaults standardUserDefaults] synchronize];
+    
+    if(self.password && ![self.password isEqualToString:FAKE_PASS_STRING]) {
+        ST::string username = ST::string([self.username cStringUsingEncoding:NSUTF8StringEncoding]);
+        ST::string password = ST::string([self.password cStringUsingEncoding:NSUTF8StringEncoding]);
+        
+        pfPasswordStore* store = pfPasswordStore::Instance();
+        if (self.rememberPassword)
+            store->SetPassword(username, password);
+        else
+            store->SetPassword(username, ST::string());
+    }
+}
+
+-(void)load {
+    NSString *serverName = [NSString stringWithSTString:GetServerDisplayName()];
+    NSDictionary *settingsDictionary = [[NSUserDefaults standardUserDefaults] dictionaryForKey:serverName];
+    self.username = [settingsDictionary objectForKey:@"LastAccountName"];
+    self.rememberPassword = [[settingsDictionary objectForKey:@"RememberPassword"] boolValue];
+    
+    if(self.rememberPassword) {
+        pfPasswordStore* store = pfPasswordStore::Instance();
+        ST::string username = [self.username stString];
+        ST::string password = store->GetPassword(username);
+        self.password = [NSString stringWithSTString:password];
+    }
+}
+
+-(void)storeHash:(ShaDigest &)namePassHash {
+    //  Hash username and password before sending over the 'net.
```

Comment:
Yeah - part of the reason I'm looking forward to feedback by @Hoikas and @dpogue is that this might happen quite a bit. I don't think we need to solve the problem now, but it would be good to start thinking about how to clean this up.